File: Repositories/go/go-microservices-grpc/services/geo/src/geo.go
package src

import (
	"encoding/json"
	"fmt"
	geo "github.com/harlow/go-micro-services/geo/proto"
	"log"
	"math"
	"net"

	"github.com/harlow/go-micro-services/geo/data"
	"golang.org/x/net/context"
	"google.golang.org/grpc"
)

const (
	maxSearchRadius = 10
	earthRadiusKm   = 6371
)

// point represents a hotel's geographic location on map
type point struct {
	Pid  string  `json:"hotelId"`
	Plat float64 `json:"lat"`
	Plon float64 `json:"lon"`
}

// Implement Point interface
func (p *point) Lat() float64 { return p.Plat }
func (p *point) Lon() float64 { return p.Plon }
func (p *point) Id() string   { return p.Pid }

// New returns a new server
func New() *Geo {
	return &Geo{
		points: loadPoints("data/geo.json"),
	}
}

// Server implements the geo service
type Geo struct {
	points []*point
}

// Run starts the server
func (s *Geo) Run(port int) error {
	srv := grpc.NewServer()
	geo.RegisterGeoServer(srv, s)

	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		return fmt.Errorf("failed to listen: %v", err)
	}

	return srv.Serve(lis)
}

// Haversine formula to calculate the distance between two points on the earth
func haversineDistance(lat1, lon1, lat2, lon2 float64) float64 {
	// Convert latitude and longitude from degrees to radians
	lat1Rad, lon1Rad := lat1*math.Pi/180, lon1*math.Pi/180
	lat2Rad, lon2Rad := lat2*math.Pi/180, lon2*math.Pi/180

	// Haversine formula
	dLat := lat2Rad - lat1Rad
	dLon := lon2Rad - lon1Rad
	a := math.Sin(dLat/2)*math.Sin(dLat/2) + math.Cos(lat1Rad)*math.Cos(lat2Rad)*math.Sin(dLon/2)*math.Sin(dLon/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))

	return earthRadiusKm * c
}

// Nearby returns all hotels within a given distance.
func (s *Geo) Nearby(ctx context.Context, req *geo.Request) (*geo.Result, error) {
	var nearbyHotels []string
	for _, point := range s.points {
		distance := haversineDistance(float64(req.Lat), float64(req.Lon), point.Plat, point.Plon)
		if distance <= maxSearchRadius {
			nearbyHotels = append(nearbyHotels, point.Pid)
		}
	}

	return &geo.Result{HotelIds: nearbyHotels}, nil
}

func loadPoints(path string) []*point {
	var (
		file   = data.MustAsset(path)
		points []*point
	)

	if err := json.Unmarshal(file, &points); err != nil {
		log.Fatalf("Failed to load hotels: %v", err)
	}

	return points
}

File: Repositories/python/python-microservices-grpc/services/geo/src/geo.py
import math
import time
from concurrent import futures
from dataclasses import dataclass

import grpc
from data.data_load_module import load_data, build_hotel_geo_data, hotel_geo_data
from proto import geo_pb2, geo_pb2_grpc

GEO_SERVICE_ADDRESS = "[::]:8080"
EARTH_RADIUS_KM = 6371.0
MAX_SEARCH_RADIUS_KM = 10  # limit to 10 km


@dataclass
class Point:
    point_latitude: float
    point_longitude: float


def haversine_distance(coord1, coord2):
    lat1, lon1 = map(math.radians, [coord1.point_latitude, coord1.point_longitude])
    lat2, lon2 = map(math.radians, [coord2.point_latitude, coord2.point_longitude])

    dlat = lat2 - lat1
    dlon = lon2 - lon1

    a = math.sin(dlat / 2) * math.sin(dlat / 2) + math.cos(lat1) * math.cos(
        lat2
    ) * math.sin(dlon / 2) * math.sin(dlon / 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    return EARTH_RADIUS_KM * c


def find_nearby_hotels(point, radius):
    nearby_hotels = []

    for hotel_id, hotel_info in hotel_geo_data.items():
        hotel_coord = Point(
            point_latitude=hotel_info["lat"], point_longitude=hotel_info["lon"]
        )
        distance = haversine_distance(point, hotel_coord)

        if distance <= radius:
            nearby_hotels.append(hotel_id)

    return nearby_hotels


class GeoServicer(geo_pb2_grpc.GeoServicer):
    def Nearby(self, request, context):
        point = Point(point_latitude=request.lat, point_longitude=request.lon)
        hotel_ids = find_nearby_hotels(point, MAX_SEARCH_RADIUS_KM)
        return geo_pb2.Result(hotelIds=hotel_ids)


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    geo_pb2_grpc.add_GeoServicer_to_server(GeoServicer(), server)
    server.add_insecure_port(GEO_SERVICE_ADDRESS)
    load_data()
    build_hotel_geo_data()
    server.start()
    try:
        while True:
            time.sleep(86400)
    except KeyboardInterrupt:
        server.stop(0)
